### PR TITLE
update sam writer factory to cope with CRAM file

### PR DESCRIPTION
**Pull request by ChristinaXu2017**
# Description

Our adamajava tools create SAM/BAM outputs by calling org.qcmg.picard.SAMOrBAMWriterFactory, but it won't create CRAM output.  Here we update the code but not API, it will make CRAM output if the output file name ending with .cram. The detailed changes are:

- rename SAMOrBAMWriterFactory to SAMWriterFactory
- create a new constructor SAMWriterFactory(..., File reference ), it takes the FASTA reference file and also allows null.
- change method closeWriter() to renameIndex(); Because the writer can automatically close by try clause but our main purpose is to rename the BAM index file from aln.bai to aln.bam.bai and the CRAM index from aln.cram.bai to aln.cram.crai. 
- tidy up code:
  -  delete some commented out code; 
  -  move renameIndex(File output) from RenameFile.java to BAMFileUtils.java
  -  replace deprecated SAMTagUtils with SAMTag;
  -  override FastaRecord.toString(), so the record can output to file. 
- use try clause to open/close file reader or writer

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Existing unit tests are updated and new unit tests for CRAM file is added

# Are WDL Updates Required?

No

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: qpicard/src/org/qcmg/picard/SAMWriterFactory.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	
+	public SAMFileWriter getWriter(){		
+		return writer;
+	}
+	/**
+	 * picard creates index aln.bai for aln.bam; and aln.cram.bai for aln.cram.
+	 * Here we rename aln.bai to aln.bam.bai; and lan.cram.bai to aln.cram.crai.
+	 */
+	public void renameIndex(){
```
Comment (holmeso) [inline qpicard/src/org/qcmg/picard/SAMWriterFactory.java:117]: I think this method should be renamed to something like `closeWriterAndRenameIndex` thus explicitly depicting what the method does. The current name is misleading in that regard.

**Review by holmeso — CHANGES_REQUESTED** [review]
Is JP happy with the index name changing to` .bam.bai `? I was under the impression that the GATK pipeline generated `.bai` rather than `.bam.bai `and I though JP wanted us to be consistent with the broadies

**Comment by ChristinaXu2017** [comment]
thanks for spotting it,  this function only works for renaming, the writer will automatically close by the try block.

**Comment by ChristinaXu2017** [comment]
The java HTSJDK library creates <lan>.bai for bam index;  <lan>.cram.bai for cram index. The SAMTOOLs c package creates <lan>.bam.bai for bam index;  <lan>.cram.crai for cram index. 

It is arguable whether we should follow HTSJDK or SAMTOOLs. I checked  "gatk-workflows:five-dollar-genome-analysis-pipeline". It calls "picard.jar SortSam" to create <aln>.bai  during "ToBam.UnmappedBamToAlignedBam";   however then, it calls SAMTOOLs to index the cram file and then create <aln>.cram.crai.

File: qpicard/src/org/qcmg/picard/util/BAMFileUtils.java [hunk 2]
```diff
@@ -46,4 +52,47 @@ public static List<String> getContigsFromHeader(SAMFileHeader header) {
 				.collect(Collectors.toList());
 	}
 
+	/**
+	 * picard index name created by replacing .bam to .bai. But we want to have index name end with .bam.bai
+	 * @param bamFile
+	 * @throws Exception 
+	 */	
+	public static void renameBamIndex(File output) throws IOException{
```
Comment (holmeso) [inline qpicard/src/org/qcmg/picard/util/BAMFileUtils.java:60]: It's usually a good idea to check the input for `public static` methods. What will happen here if output is null?

File: qpicard/src/org/qcmg/picard/util/BAMFileUtils.java [hunk 3]
```diff
@@ -46,4 +52,47 @@ public static List<String> getContigsFromHeader(SAMFileHeader header) {
...
+		//rename files
+		Files.move(indexpicard, indexqcmg, StandardCopyOption.REPLACE_EXISTING);				
+	}
+	/**
+	 * picard index name created by adding .bai to .cram. But we want to have index name end with .cram.crai
+	 * @param cramFile
+	 * @throws Exception 
+	 */	
+	
+	public static void renameCramIndex(File output) throws IOException{
```
Comment (holmeso) [inline qpicard/src/org/qcmg/picard/util/BAMFileUtils.java:81]: It's usually a good idea to check the input for `public static` methods. What will happen here if output is null?

File: qmule/src/org/qcmg/qmule/BAMCompress.java [hunk 4]
```diff
@@ -21,7 +21,7 @@
 import org.qcmg.common.log.QLogger;
 import org.qcmg.common.log.QLoggerFactory;
 import org.qcmg.picard.SAMFileReaderFactory;
-import org.qcmg.picard.SAMOrBAMWriterFactory;
+import org.qcmg.picard.SAMWriterFactory;
```
Comment (holmeso) [inline qmule/src/org/qcmg/qmule/BAMCompress.java:24]: Is this import used?

File: qcnv/test/org/qcmg/cnv/TestFiles.java [hunk 5]
```diff
@@ -77,12 +77,10 @@ private static void Txt2BAM(File SAM, File BAM) throws IOException{
 		SAMFileHeader header = reader.getFileHeader().clone();
 		header.setSortOrder(SAMFileHeader.SortOrder.coordinate);
 		
-		SAMOrBAMWriterFactory factory = new SAMOrBAMWriterFactory(header, false, BAM, true);
-		SAMFileWriter writer = factory.getWriter();
-		for( SAMRecord record : reader)
-			writer.addAlignment(record);	
-			
-		reader.close();
```
Comment (holmeso) [inline qcnv/test/org/qcmg/cnv/TestFiles.java:85]: why is the reader no longer being closed?

File: qbamfix/src/org/qcmg/qbamfix/FixBAM.java [hunk 6]
```diff
@@ -187,48 +187,39 @@ HashMap<String, Integer> checkMate(List<SAMRecord> reads, File inBAM)throws IOEx
...
-	    			writer.addAlignment(record);	
-	    			NumofOutput ++;
-	    		} else {
-	    			badReads.add(record);
-	    		}
-	    		NumofInput ++;	
+        boolean ok2add;		
+		
+		try(SamReader reader = SAMFileReaderFactory.createSAMFileReader(input, null,  validation);) {
+			SAMWriterFactory factory = 	reader.getFileHeader().getSortOrder().equals(header.getSortOrder())?
```
Comment (holmeso) [inline qbamfix/src/org/qcmg/qbamfix/FixBAM.java:197]: could be simplified to:
`SAMWriterFactory factory = new SAMWriterFactory(header, reader.getFileHeader().getSortOrder().equals(header.getSortOrder()), output, tmpDir, 2000000, true )`

File: qpicard/src/org/qcmg/picard/util/BAMFileUtils.java (hunk 2 above)
Comment (ChristinaXu2017) [reply qpicard/src/org/qcmg/picard/util/BAMFileUtils.java:60]: thanks. I will check the input.

File: qmule/src/org/qcmg/qmule/BAMCompress.java (hunk 4 above)
Comment (ChristinaXu2017) [reply qmule/src/org/qcmg/qmule/BAMCompress.java:24]: thanks, fixed

File: qcnv/test/org/qcmg/cnv/TestFiles.java (hunk 5 above)
Comment (ChristinaXu2017) [reply qcnv/test/org/qcmg/cnv/TestFiles.java:85]: thanks, fixed

File: qbamfix/src/org/qcmg/qbamfix/FixBAM.java (hunk 6 above)
Comment (ChristinaXu2017) [reply qbamfix/src/org/qcmg/qbamfix/FixBAM.java:197]: yes, simplified.

File: qpicard/src/org/qcmg/picard/util/BAMFileUtils.java [hunk 7]
```diff
@@ -57,8 +57,11 @@ public static List<String> getContigsFromHeader(SAMFileHeader header) {
 	 * @param bamFile
 	 * @throws Exception 
 	 */	
-	public static void renameBamIndex(File output) throws IOException{
-		String path = output.getPath();
+	public static void renameBamIndex(File input) throws IOException{
+		
+		if( input == null ) throw new IOException("we can't rename a NULL file!");
```
Comment (holmeso) [inline qpicard/src/org/qcmg/picard/util/BAMFileUtils.java:62]: IllegalArgumentException rather that IOException

File: qpicard/src/org/qcmg/picard/util/BAMFileUtils.java [hunk 8]
```diff
@@ -70,16 +73,20 @@ public static void renameBamIndex(File output) throws IOException{
...
 	 * picard index name created by adding .bai to .cram. But we want to have index name end with .cram.crai
 	 * @param cramFile
 	 * @throws Exception 
 	 */	
 	
-	public static void renameCramIndex(File output) throws IOException{
-		String path = output.getPath();
+	public static void renameCramIndex(File input) throws IOException{
+		
+		if( input == null ) throw new IOException("we can't rename a NULL file!");
```
Comment (holmeso) [inline qpicard/src/org/qcmg/picard/util/BAMFileUtils.java:87]: IllegalArgumentException rather that IOException

File: qpicard/test/org/qcmg/picard/util/BAMFileUtilsTest.java [hunk 9]
```diff
@@ -73,16 +73,25 @@ public void renameIndexTest() throws IOException {
...
 		Path dir = Files.createTempDirectory("java-test");
 		File bam = new File(dir.resolve("test.bam").toString());
 		// create the bam but no index
 		bam.createNewFile();
 		// expect IOException
 		BAMFileUtils.renameBamIndex(bam);
+		
+	}
+	
+	@Test(expected = IOException.class)
```
Comment (holmeso) [inline qpicard/test/org/qcmg/picard/util/BAMFileUtilsTest.java:90]: IllegalArgumentException rather that IOException

**Review by holmeso — APPROVED** [review]

File: qpicard/src/org/qcmg/picard/util/BAMFileUtils.java (hunk 8 above)
Comment (ChristinaXu2017) [reply qpicard/src/org/qcmg/picard/util/BAMFileUtils.java:87]: yes, I agree It is an Argument Exception but here this argument is also IO related.  Actually, it is a NullPointerException but we just want to handle it and throw more related error message.

Comment (holmeso) [reply qpicard/src/org/qcmg/picard/util/BAMFileUtils.java:87]: Check out the javadocs: https://docs.oracle.com/javase/8/docs/api/java/lang/IllegalArgumentException.html

**Pull request merged**